packages/fast-components-react-base/src/toolbar/toolbar.tsx

```diff
@@ -0,0 +1,355 @@
+            itemPath = focusableItems[0].getAttribute(Toolbar.toolbarItemAttributeName);
+        }
+
+        this.setState({
+            focusItemPath: itemPath,
+        });
+    };
+
+    /**
+     * Handle the keydown event
+     */
+    private handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
+        // increment focused widget on arrow keys
+        if (this.props.orientation === Orientation.horizontal) {
+            switch (e.keyCode) {
+                case keyCodeArrowRight:
+                    e.preventDefault();
+                    this.incrementFocus(this.state.focusItemPath, 1);
+                    break;
+
+                case keyCodeArrowLeft:
+                    e.preventDefault();
+                    this.incrementFocus(this.state.focusItemPath, -1);
+                    break;
+            }
+        } else {
+            switch (e.keyCode) {
+                case keyCodeArrowDown:
+                    e.preventDefault();
+                    this.incrementFocus(this.state.focusItemPath, 1);
+                    break;
+
+                case keyCodeArrowUp:
+                    e.preventDefault();
+                    this.incrementFocus(this.state.focusItemPath, -1);
+                    break;
+            }
+        }
+
+        // shortcuts
+        let focusableWidgets: HTMLElement[] = null;
+        switch (e.keyCode) {
+            case keyCodeHome:
+                e.preventDefault();
+                focusableWidgets = this.getFocusableItems();
+                if (focusableWidgets.length === 0) {
+                    return;
+                }
+                focusableWidgets[0].focus();
+                break;
+
+            case keyCodeEnd:
+                e.preventDefault();
+                focusableWidgets = this.getFocusableItems();
+                if (focusableWidgets.length === 0) {
+                    return;
+                }
+                focusableWidgets[focusableWidgets.length - 1].focus();
+                break;
+        }
+    };
+
+    /**
+     * Respond to focus changes within the component (on blur/focus capture)
+     * Essentially we want to validate that the default focus item is valid
+     */
+    private updateFocusItemPath = (e: React.FocusEvent<HTMLElement>): void => {
+        if (e.target.hasAttribute(Toolbar.toolbarItemAttributeName)) {
+            this.setState({
+                focusItemPath: e.target.getAttribute(Toolbar.toolbarItemAttributeName),
+            });
+            return;
+        }
+
+        this.validateCurrentFocusItem();
+    };
+
+    /**
+     *  Increments focus relative to the item with the provided path
+     */
+    private incrementFocus = (focusPath: string, adjustment: number): void => {
+        const focusableItems: HTMLElement[] = this.getFocusableItems();
+
+        if (focusableItems.length === 0) {
+            return;
+        }
+
+        let targetItemIndex: number = -1;
```

Comment:
Just as an FYI, there is a "findIndex" method on the array you can use: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex

This works too, but if you do want to make this more concise you could use it.

Reply:
Yes, much nicer.  Thx